src/util/bip152.rs

```diff
@@ -0,0 +1,476 @@
+    #[inline]
+    fn consensus_decode<D: io::Read + ?Sized>(mut d: &mut D) -> Result<PrefilledTransaction, encode::Error> {
+        let idx = VarInt::consensus_decode(&mut d)?.0;
+        if idx > u16::max_value() as u64 {
+            //TODO use TryInto instead when MSRV exceeds 1.34
+            return Err(encode::Error::ParseFailed("BIP152 prefilled tx index out of bounds"));
+        }
+        let tx = Transaction::consensus_decode(&mut d)?;
+        Ok(PrefilledTransaction { idx: idx as u16, tx })
+    }
+}
+
+/// Short transaction IDs are used to represent a transaction without sending a full 256-bit hash.
+#[derive(PartialEq, Eq, Clone, Copy, Hash, Default, PartialOrd, Ord)]
+pub struct ShortId(pub [u8; 6]);
+
+impl ShortId {
+    /// Calculate the SipHash24 keys used to calculate short IDs.
+    pub fn calculate_siphash_keys(header: &BlockHeader, nonce: u64) -> (u64, u64) {
+        // 1. single-SHA256 hashing the block header with the nonce appended (in little-endian)
+        let h = {
+            let mut engine = sha256::Hash::engine();
+            header.consensus_encode(&mut engine).expect("engines don't error");
+            nonce.consensus_encode(&mut engine).expect("engines don't error");
+            sha256::Hash::from_engine(engine)
+        };
+
+        // 2. Running SipHash-2-4 with the input being the transaction ID and the keys (k0/k1)
+        // set to the first two little-endian 64-bit integers from the above hash, respectively.
+        (endian::slice_to_u64_le(&h[0..8]), endian::slice_to_u64_le(&h[8..16]))
+    }
+
+    /// Calculate the short ID with the given (w)txid and using the provided SipHash keys.
+    pub fn with_siphash_keys<T: AsRef<[u8]>>(txid: &T, siphash_keys: (u64, u64)) -> ShortId {
+        // 2. Running SipHash-2-4 with the input being the transaction ID and the keys (k0/k1)
+        // set to the first two little-endian 64-bit integers from the above hash, respectively.
+        let hash = siphash24::Hash::hash_with_keys(siphash_keys.0, siphash_keys.1, txid.as_ref());
+
+        // 3. Dropping the 2 most significant bytes from the SipHash output to make it 6 bytes.
+        let mut id = ShortId([0; 6]);
+        id.0.copy_from_slice(&hash[0..6]);
+        id
+    }
+}
+
+impl hex::FromHex for ShortId {
+    fn from_byte_iter<I>(iter: I) -> Result<Self, hex::Error>
+    where
+        I: Iterator<Item = Result<u8, hex::Error>> + ExactSizeIterator + DoubleEndedIterator,
+    {
+        Ok(ShortId(hex::FromHex::from_byte_iter(iter)?))
+    }
+}
+
+impl fmt::LowerHex for ShortId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        hex::format_hex(&self.0[..], f)
+    }
+}
+
+impl fmt::Display for ShortId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::LowerHex::fmt(self, f)
+    }
+}
+
+impl fmt::Debug for ShortId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
```

Comment:
These can be replaced with `impl_bytes_newtype!(ShortId, 6);`

Reply:
Note, doing so would add three more implementations as well, `FromStr`, `Serialize`, and `Deserialize`. Not sure right now if that's ok so flagging it.

Reply:
I think the extra impls are probably good ideas.